lilcom/lilcommodule.c

```diff
@@ -1,118 +1,193 @@
+                        just in dimension
+  
+    @param [out] output   The 8-bit compresed data:  a numpy
+                        array passed to the function, which should
+                        be the proper sub-array of the output to paste
+                        the output array.
  */
-static PyObject * compress(PyObject * self, PyObject * args)
+static PyObject * compress16i_8i(PyObject * self, PyObject * args, PyObject * keywds)
 {
-
-  PyObject *signalObj; // The input signal, passed as a numpy array.
-  if (!PyArg_ParseTuple(args, "O", &signalObj)) { //Parsing the input of the python code
-      Py_RETURN_FALSE;
-  }
-  int n_dims = PyArray_NDIM(signalObj); // Number of dimensions in the numpy array
-  int stride; // The number of integers between to consecutive samples.
-
-  if (n_dims == 1){
-    stride = 1;
-  } // One dimensional array does't have any problem in stride.
-
-  // Finding the shape of array
-  int n_samples = (int)PyArray_DIM(signalObj, 0);
-  if (n_dims != 1)
-    stride = (int)PyArray_DIM(signalObj, 1);
-
-  int **singal_stream = (int **)(PyArray_DATA(signalObj));
-
-  // Conversion to int16_t
-  int16_t *input = malloc(sizeof(int16_t) * n_samples * stride); // alocating a linear array of considered type.
-  for (int i = 0 ; i < n_samples * stride; i++){
-    input[i] = singal_stream[i];
-  }
-
-  int8_t * output;
-  int output_stride = 1;
-  int lpc_order = 1;
-
-  // FUNCTION CALL HERE!
-  printf("function call!\n");
-  lilcom_compress(n_samples, input, stride, output, output_stride, lpc_order);
-
-
-  Py_RETURN_TRUE;
+  /* Defining all variables */
+  PyObject *signal_input; // The input signal, passed as a numpy array.
+  PyObject *signal_output; // The output signal, passed as a numpy array.
+  int n_dims = 0; // Number of dimensions of the given numpy array
+  int input_stride; // The number of integers between to consecutive samples
+  int output_stride; // The number of integers between to consecutive samples in output
+  int n_samples = 0; // number of samples given in the numpy array
+  // INPUT IS DEFINED INSIDE IF CONDITIONS
+  int8_t *output; // The one dimensional vectorized output array which will be modified by the core function
+  int16_t *input; // The one dimensional vectorized input array which will be given to the core function
+  int lpc_order = 5; // LPC Order defined in the core function (more information -> lilcom.h)
+  int conversion_exponent = 0; // Conversion Exponent defined in the core function (more information -> lilcom.h)
+
+  /* Reading and information - extracting for input data 
+    From the python function there are two numpy arrays and an intger (optional) LPC_order 
+    passed to this madule. Following part will parse the set of variables and store them in corresponding
+    objects.
+  */
+  static char *kwlist[] = {"X", "Y" , "lpc_order", "conversion_exponent", NULL}; //definition of keywords received in the function call from python
```

Comment:
Not sure what X and Y are?

Reply:
Also, for this function to handle arrays with arbitrary numbers of axes, the easiest way would be to have the actual implementation as a recursive function.   The interface could have things like raw int16_t and int8_t pointers and the strides and dims specified as pointers to ints.   To avoid having to malloc, you can just require that the num-axes be less than 10 and use local arrays (make sure to return false if it is more, though, to avoid security issues).

Reply:
/**
  Recursive internal implementation of lilcom_compress_16i_8i.
  @param [in] ndim  The number of dimensions in the arrays (which must be the same, and
                                 must be >= 1).
  @param [in] dim   The dimension that this function is to process.    If equal to ndim-1, it
                                actually does the compression (this dimension corresponds to the time axis).
                                For values 0 ... ndim-2, this function will loop and recurse.
  @param [in] input_data . Raw pointer to the input data; will have been obtained from
                               calling input_array and possibly shifted by outer calls of this function
  @param [out] output_data  Raw pointer to the output data; obtained from output_array
                               and possibly shifted, as for input_data.
   @param [in] input_array  Must point to a NumPy array representing the input.  Used
                               for its dimension and stride information
   @param [in] output_array  Must point to a NumPy array representing the output data.  Used
                               for its dimension and stride information
    @param [in] lpc_order . ....
    @param [in] conversion_exponent ...
   @return   Returns 0 on success, 1 on failure (e.g. if there was a mismatch in dimensions, 
                 or a call to lilcom_compress failed
*/
int compress16i_8i_internal(int ndim, int dim, const int16_t *input_data, int8 *output_data,
                                               PyObject *input_array, PyObject *output_array, int lpc_order,
                                               int conversion_exponent) {
}

(PyObject * self, PyObject * args, PyObject * keywds)

Reply:
`X` is the input array, and `Y` is the output array.

Reply:
I used this notation to make it like scikit

Reply:
For using multi-dimensional arrays I think the best way is to do this in wrapper. (Also if you want it to be written in c, it's ok, I will do it in the next step) As long as numpy itself returns subarrays along axis, we can just call the function with proper strides. I just did this branch to have basic and simple element functions. And dealing with the more complex situations  will be done using these building blocks.

Reply:
OK... X and Y didn't seem to match with the names you put in the documentation.  I'd prefer "input" and "output" or something like that, it's more consistent with the other code.
The reason I was suggesting to have the loop in "C" was for efficiency reasons; loops in python will be slow, and I want this to deal efficiently with short sequences.  Hm.. maybe I'll take your branch and do some work on it myself, actually; I might learn something.